shared_model/interfaces/hashable.hpp

```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_HASHABLE_HPP
+#define IROHA_HASHABLE_HPP
+
+#include "interfaces/common_objects/hash.hpp"
+#include "interfaces/primitive.hpp"
+
+namespace shared_model {
+  namespace interface {
+    template <typename ModelType, typename OldModel>
+    class Hashable : public Primitive<ModelType, OldModel> {
+     public:
+      /// Type of hash
+      using HashType = Hash;
+
+      /**
+       * @return hash of object.
+       * Equality of hashes means equality of objects.
+       */
+      virtual const HashType &hash() const = 0;
+
+      /**
+       * Overriding operator== with equality hash semantics
+       * @param rhs - another model object
+       * @return true, if hashes are equal, false otherwise
+       */
+      bool operator==(const ModelType &rhs) const override {
```

Comment:
Not really sure that it's a good idea.

If I properly understand hierarchy some type should be derived from hashable (or its child)
Then how can you intend to provide method on comparing objects? `hash(a)==hash(b)` doesn't mean `a==b`

Reply:
By the way if that method is removed then it's possible remove class' ancestor

Reply:
In general case, it may be not true. If your business objects should change hash equality behavior, they can do it. This overring useful for a transaction, block, and query at all.

Reply:
>By the way if that method is removed then it's possible remove class' ancestor

No, because signable and hashable are different types of objects. Signable: tx, query, block. Hashable(but not Signable): concrete command, concrete query etc.
In additional Primitive by itself also useful. For example query and tx responses not required to be hashable by itself.